Sparse Tensor: request for reviews

eric-haibin-lin opened the pull request:
Following please find a summary of changes made for sparse tensor feature on CPU. 

Note: this PR contains duplicate code in PR #7015 

# Frontend Interfaces

## NDArray Class Hierarchy 

-> NDArrayBase (root)
-> NDArray
-> SparseNDArray
-> RowSparseNDArray, CSRNDArray

## Sparse Storage Types Overview

Two storage formats are supported: `row_sparse` and `csr`. 

- [`row_sparse`](https://github.com/dmlc/mxnet/blob/sparse/python/mxnet/ndarray/sparse_ndarray.py#L378) tensors, typically used for sparse gradient updates (similar to [`tf.IndexedSlice`](https://www.tensorflow.org/api_docs/python/tf/IndexedSlices)) 

- [`csr`](https://github.com/dmlc/mxnet/blob/sparse/python/mxnet/ndarray/sparse_ndarray.py#L331) tensors, with the [standard csr storage format](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_.28CSR.2C_CRS_or_Yale_format.29). 

## Functionality Overview
### operators with sparse inputs
A list of operators in this PR include:

* Arithmetics: `elemwise_add(row_sparse)`, `dot(csr, row_sparse)`, `dot(csr, dense)`
* Conversion: `cast_storage(csr<->dense)`, `cast_storage(row_sparse<->dense)`
* Indexing: [`sparse_retain(row_sparse)`](https://github.com/dmlc/mxnet/blob/sparse/src/operator/tensor/indexing_op.cc#L232), `slice(csr)`
* Creation: `zeros(row_sparse)`, `zeros(csr)`


### iterators and IO utils functions
- `NDArrayIter` with `CSRNDArray`
- `LibsvmIter` which reads files in libsvm format and produces `CSRNDArray`
- load/save `RowSparseNDArray`/`CSRNDArray`

### kvstore and optimizer with sparse gradient updates
- `sgd_update(row_sparse)` and `sgd_mom_update(row_sparse)` will update rows of the weight where the gradient of the row contains nonzeros.
- `kv.push` with row_sparse gradient, which only sends over the nonzero gradients over the network
- `kv.row_sparse_pull` with row_ids, which only pulls back a RowSparseNDArray with specified row_id over the network

### unsupported functionality 
There's a list of basic functionality supported in NDArray, but is either not supported or inefficient in SparseNDArray:
- get/set an element in CSRNDArray/RowSparseNDArray by index (not supported)
- `reshape` in CSRNDArray/RowSparseNDArray (not supported)
- `_slice`, `_at` in CSRNDArray/RowSparseNDArray (not supported)
- inplace add/sub/mul/div/mod, `broadcast_to` in CSRNDArray/RowSparseNDArray (not efficient)

Please refer to the TODO list for a list of ongoing/planned operators.

# Backend Interfaces

## NDArray

NDArray uses `storage_type` to determine what storage format to use to represent a tensor.

```cpp
enum NDArrayStorageType {
  kUndefinedStorage = -1,  // undefined storage
  kDefaultStorage,         // dense
  kRowSparseStorage,       // row sparse
  kCSRStorage,             // csr
};
```

Some of the important interfaces added to NDArray are:

```cpp
NDArray {
  // constructor
  NDArray(const NDArrayStorageType stype, const TShape &shape, Context ctx,
          bool delay_alloc = true, int dtype = mshadow::default_type_flag,
          std::vector<int> aux_types = {}, std::vector<TShape> aux_shapes = {},
          TShape storage_shape = TShape(mshadow::Shape1(0)))
  // memory allocation
  void CheckAndAllocData(const TShape &storage_shape) const; 
  void CheckAndAllocAuxData(size_t i, const TShape &aux_shape) const;

  struct Chunk {
    Storage::Handle shandle;
    // storage handles for aux data (e.g. indptr, indices)
    std::vector<Storage::Handle> aux_handles;
    NDArrayStorageType storage_type;
    std::vector<int> aux_types;
    TShape storage_shape;
    std::vector<TShape> aux_shapes;
    void CheckAndAllocData(const TShape &shape, int dtype);
    void CheckAndAllocAuxData(size_t i, const TShape &shape);
  };
...

 void CopyFromTo(const NDArray &from, NDArray *to) {
    // if storage_types of from & to don't match, convert to destination storage type
    // perform copy...
 }

```

## Executor and NNVM

### operator interface

To deal with tensors of non-default storage, operators need to register with the following interface:
```cpp
using FComputeEx = std::function<void (
        const nnvm::NodeAttrs& attrs,
        const OpContext& ctx,
        const std::vector<NDArray>& inputs,
        const std::vector<OpReqType>& req,
        const std::vector<NDArray>& outputs)>;
```

Since memory usage for sparse tensor cannot be calculated during graph binding phase, operator is expected to use `CheckAndAllocAuxData` and `CheckAndAllocData` to perform memory allocation for result tensors at runtime. 

### infer_storage_type pass 

Besides `FInferShape` and `FInferType`, an operator needs to register `FInferStorageType` if it involves non-default storage_type. `ctx` could be used to infer if MKL experimental memory should be used.  The attribute `__storage_type__` of `mx.sym.Variable` is read for this pass.

```cpp
using FInferStorageType = std::function<bool (
        const NodeAttrs& attrs,
        const Context& ctx,
        std::vector<int>* in_attrs,
        std::vector<int>* out_attrs)>;
```

When binding a graph, `InferStorageType` pass will be applied to infer graph entries with unknown storage_type. The pass is implemented in MXNet instead of NNVM. `infer_type` and `infer_shape` passes are both moved from NNVM to MXNet.

### plan_memory pass

For node entries of non-default storage_type, they're annotated with `kDynamicStorageID`; memory are neither requested nor released. No memory sharing optimization is applied. 

### memory allocation and sharing

For node entries of `kDefaultStorage`(i.e. dense storage), memory allocation and sharing is done as usual. For node entries of non-default storage_type, memory allocation is delayed to runtime. No memory sharing is applied to these entries even when `shared_exec` is provided.

### executor dispatch and storage fallback

For each node in graph, if it involves non-default storage in either its inputs or outputs, `FComputeExExecutor` will be created if `FComputeEx` attribute is registered. 

If `FComputeEx` is not registered, an `FComputeExcutor`/`FStatefulComputeExcutor` will be created, inputs with non-default storage will be converted to NDArrays with default (i.e. dense) storage; the outputs will be converted back to non-default storage if necessary. The converted NDArrays are temporary and will be released when there's no reference to them.

## kvstore

Sparse gradient update requires a parameter to be initialized as row_sparse format in kvstore. If an optimizer is registered with `FComputeEx`, then the update may only involve the rows with non-zero gradients. Gradients are sent in `row_sparse` format to reduce the stress on network bandwidth. 

For distributed training, `kRowSparsePushPull` mode is introduced so that row ids of the row_sparse gradients are encoded as keys by kvstore workers. When the kvstore server receives the message, it constructs a row_sparse gradient NDArray based on the row ids decoded from it and applies the optimizer to the weight. The memory for parameters are pre-allocated on parameter servers, instead of storing weight row by row.


# TODO list
## Operators
- elemwise binary & unary operators (row_sparse): done, to be merged in another PR
  negative, square, abs, sqrt, round, ceil, floor, fix, elemwise_add, elemwise_sub, elemwise_div, elemwise_mul, _maximum, _minimum, _mul_scalar
- elemwise binary & unary operators (csr): ongoing
  negative, square, broadcast_mul, _mul_scalar
- inplace broadcast_add, broadcast_sub, _mul_scalar operators
- broadcast operators (csr)
- broadcast operators (row_sparse)
- elemwise operators(csr)
- `_sync_copy_from` with `scipy.sparse.csr_matrix` input
- `asscipy` (csr)


## kvstore and optimizer 
- lazy initialization - postpone parameter initialization to kvstore server
- storing `row_sparse` weight by rows to support weights of > 2 billion rows. This avoid pre-allocating memory for row_sparse weight and reduces memory consumption at kvstore servers.
- other optimizers for sparse gradient update (e.g. adam)

## List of important files 

#### NDArray related files:
- ndarray.h
- c_api.h
- ndarray.cc
- common/utils.h

#### Executor related files:
- c_api_executor.h
- attach_op_exec_pass.cc
- graph_attr_types.h
- op_attr_types.h
- executor.h
- graph_executor.h
- graph_executor.cc
- infer_graph_attr_pass.cc
- inplace_addto_detect_pass.cc

#### Kvstore related files:
- comm.h
- kvstore_dist.h
- kvstore_dist_server.h
- kvstore_dist_local.h
(pending PR #7015)

piiswrong commented on python/mxnet/ndarray/__init__.py:
```diff
@@ -0,0 +1,12 @@
+from . import op
+from .op import CachedOp
+from .ndarray import NDArray, array, concatenate, _DTYPE_NP_TO_MX, _DTYPE_MX_TO_NP
+from .ndarray import empty, ones, add, arange, divide, equal, full, greater, greater_equal, imdecode
```
It's ok to use wild card import

eric-haibin-lin replied:
will fix this in next PR

eric-haibin-lin commented on src/operator/tensor/elemwise_unary_op.h:
```diff
@@ -91,7 +92,11 @@ void IdentityComputeRsp(const nnvm::NodeAttrs& attrs,
+    FillZerosRspImpl(s, &out);
+    return;
+  }
   TShape shape = input.aux_shape(rowsparse::kIdx);
```
@cjolivier01 maybe you fixed it already but just in case you haven't. This may lead to test failure for binary/unary ops

eric-haibin-lin commented:
@ZihengJiang @szha @anirudh2290 @stefanhenneking @madjam @cjolivier01
@tqchen @jermainewang @piiswrong @mli @pluskid @ykim362 @formath @reminisce

It's probably not feasible to review all the operators implemented, but I'd appreciate if important changes made to NDArray, executor and kvstore are reviewed. Any help is appreciated!

ZihengJiang commented on include/mxnet/c_api.h:
```diff
@@ -321,6 +353,17 @@ MXNET_DLL int MXNDArraySyncCopyToCPU(NDArrayHandle handle,
+ */
+MXNET_DLL int MXNDArraySyncCopyFromNDArray(NDArrayHandle handle_dst,
+                                           const NDArrayHandle handle_src,
+                                           const int i);
```
suggest to rename i to `index` or `type`

eric-haibin-lin replied:
@reminisce

eric-haibin-lin replied:
`i` could either be used to indicate data or aux_data, not sure if `data_idx` is a good name either.

reminisce replied:
Agree with @eric-haibin-lin. As long as it's documented clearly in the comment, it's fine to use `i`.

ZihengJiang commented on include/mxnet/ndarray.h:
```diff
@@ -439,12 +846,12 @@ class NDArray {
  * \param from the ndarray we want to copy data from
  * \param to the target ndarray
  * \param priority Priority of the action.
+ * \param alloc_output whether to allocate memory for the output ndarray
```
removed?

eric-haibin-lin replied:
Thanks for pointing this out. will remove this comment, too

jermainewang commented:
Great job! Did you consider use the storage format proposed by http://people.csail.mit.edu/fred/tensor-compiler-techreport.pdf ? It seems to me that adding more storage formats is quite tedious (e.g. CSC, BCSR, DCSR, etc.). It will be great that we can define the storage format in a hierarchical way so future extension will be much easier.

piiswrong commented on src/executor/attach_op_execs_pass.cc:
```diff
@@ -22,28 +23,39 @@ const OperatorProperty* OpPropGetOpProperty(const NodeAttrs& attrs);
+    // TODO(haibin) avoid repeating this if all inputs are already in default-storage
     op_ctx.run_ctx = rctx;
-    fcompute_(state_, op_ctx, in_data_, req, out_data_);
+    in_data_.clear();
```
is clearing everytime necessary?

piiswrong replied:
is clearing everytime necessary?

eric-haibin-lin replied:
not anymore. Changed in https://github.com/eric-haibin-lin/mxnet/pull/134

eric-haibin-lin replied:
not anymore. Changed in https://github.com/eric-haibin-lin/mxnet/pull/134

piiswrong commented on src/operator/elemwise_op_common.h:
```diff
@@ -90,6 +126,33 @@ inline bool ElemwiseType(const nnvm::NodeAttrs& attrs,
+                                           const Context& ctx,
+                                           std::vector<int> *in_attrs,
+                                           std::vector<int> *out_attrs) {
+  // TODO(junwu): add ctx info into storage inference logic
```
Why put this in common?

eric-haibin-lin commented:
@jermainewang  Thanks for the great reference! I can see how the proposed storage format in the paper could be used to express many types of common sparse storage format like CSC and CSR. As for "define the storage format in a hierarchical way", I can see the four main categories in this paper: dense, dense / dense, sparse / sparse dense / sparse, sparse. Are you suggesting we assign the NDArrayStorageType enums to some of the categories here?

eric-haibin-lin commented on include/mxnet/kvstore.h:
```diff
@@ -155,6 +156,29 @@ class KVStore {
+   * \param priority the priority of the action.
+   */
+  virtual void PullRowSparse(const std::vector<int>& str_keys,
+                             const std::vector<std::pair<NDArray*, NDArray>>& val_rowids,
```
For each value we may have different row_ids to pull. e.g.
val0.ctx = mx.gpu(0), val1.ctx = mx.gpu(1)
kv.row_sparse_pull("key", [val0, val1], [mx.nd.array([0, 1, 3]), mx.nd.array([4,5,6])]

eric-haibin-lin commented on python/mxnet/_ctypes/ndarray.py:
```diff
@@ -61,22 +70,29 @@ def _imperative_invoke(handle, ndargs, keys, vals, out):
+    # a handle's stype in _ndarray_cls
+    out_stypes = ctypes.POINTER(ctypes.c_int)()
+
+    check_call(_LIB.MXImperativeInvokeEx(
```
will do this in next PR

eric-haibin-lin commented on python/mxnet/kvstore.py:
```diff
@@ -214,11 +217,88 @@ def pull(self, key, out=None, priority=0):
+        --------
+        >>> shape = (3, 3)
+        >>> kv.init('3', mx.nd.ones(shape)._to_rsp())
+        >>> a = mx.nd.zeros(shape)
```
good catch.

eric-haibin-lin commented on python/mxnet/ndarray/sparse_ndarray.py:
```diff
@@ -0,0 +1,637 @@
+    """A RowSparseNDArray is typically used to represent a subset of a larger
+    NDArray  with `default` of shape [LARGE0, D1, .. , DN] where LARGE0 >> D0. The values
+    in indices are the indices in the first dimension of the slices that have been extracted from
+    the larger NDArray. The indices are expected to be sorted in ascending order.
```
Yes currently we put the ordered constrain here so that binary search through indices is cheap. Is it not reasonable?

eric-haibin-lin commented on python/mxnet/ndarray/sparse_ndarray.py:
```diff
@@ -0,0 +1,637 @@
+    return src, dtype
+
+
+def csr(data, indptr, indices, shape, ctx=None, dtype=None, indptr_type=None, indices_type=None):
```
actually if we want to make it longer I prefer csr_matrix instead of csr_array since it's always going to be 2D and that's how scipy call it

eric-haibin-lin replied:
Should we change mx.nd.row_sparse to mx.nd.row_sparse_array then?

jermainewang commented:
The storage format can be formalized as a vector. Each element is a pair contains "dim" and "dense/sparse".
For example:
* Row-major equals to [<0, dense>, <1, dense>]
* Col-major equals to [<1, dense>, <0, dense>]
* CSR equals to [<0, dense>, <1, sparse>]
* CSC equals to [<1, dense>, <0, sparse>]

If having more tensor dimensions (e.g. convolution),
* BCSR equals to [<0, dense>, <1, sparse>, <2, dense>, <3, dense>]

I found most variants of current sparse tensor formats can be formalized as this. The paper talks about coordinate, ellpack and dia formats in the future work, which they believe can also be supported.

In terms of implementation, we could change the enum to the vector above, define some common formats (like CSR, CSC), and define the "==" operator. This should simulate the behavior of enum type.

eric-haibin-lin commented on src/c_api/c_api_ndarray.cc:
```diff
@@ -301,10 +384,25 @@ void PushOperator(const OpStatePtr& state,
+          std::vector<OpReqType> req(output_blobs.size(), kWriteTo);
+          fcompute(state, opctx, input_blobs, req, output_blobs);
+          // cast to original storage type, if necessary
+          CastNonDefaultStorage<gpu>(ndoutputs, temp_out, opctx);
```
updated in eric-haibin-lin#134
Since storage fallback doesn't add any additional node to the graph, it only changes the way a node executes, so I think it should be fine. 
Autograd test with sparse input added in commit: https://github.com/eric-haibin-lin/mxnet/pull/134/commits/8fb3b9cdf265907ba838718638747e4edc5c802f

reminisce commented:
@jermainewang Thanks for the paper reference. I can see the mapping between our currently defined enum `kDefaultStorage`, `kRowSparseStorage`, `kCSRStorage` and vectors `[<0, dense>, <1, dense>]`, `[<0, sparse>, <1, dense>]`, `[<0, dense>, <1, sparse>]`. I have a few questions regarding the proposal though:

1. `kRowSparseStorage` mapps to `[<0, sparse>, <1, dense>]` only when the ndarray is 2D, but the enum `kRowSparseStorage` actually means **"sparse in the first dim and dense in the rest of the dims"**. It seems difficult to use a fixed-size vector to represent this generalized case for N-D tensors, where N is only known at runtime.

2. Has this formalization been proven to work for any storage format? My concern is that if we want to add a storage format that cannot be formalized in the vector representation in the future, how do we deal with the exception?

3. It is not clear to me that how [COO](https://www.tensorflow.org/api_docs/python/tf/SparseTensor) format fits in the scheme? It seems that this kind of storage format is not mentioned/considered in the paper.

jermainewang commented:
@reminisce These are actually great catches!

1. I think you are right. What I could think about is to use a integer (32/64 bits) to represent the format. Each bit means sparse or dense. In this case, we can support up to 32/64 dimension tensors, which may be enough?

2/3. I am not sure about this. I raised an issue in there project so we can discuss directly with the authors. (https://github.com/tensor-compiler/taco/issues/108)

tqchen commented:
I think the format layout by taco is interesting and we should adopt this more general scheme

eric-haibin-lin commented:
These are good discussions. A more general format will make adding more storage types easier. We should also take special formats such as MKL experimental memory into consideration since it uses StorageType, too. Also we need to evaluate the cost of interface changes and performance if we want to use a vector instead of integer for NDArraryStorageType.

eric-haibin-lin commented on python/mxnet/ndarray/sparse_ndarray.py:
```diff
@@ -0,0 +1,637 @@
+
+# pylint: disable=abstract-method
+class RowSparseNDArray(SparseNDArray):
+    """A RowSparseNDArray is typically used to represent a subset of a larger
```
I copied the one from tf.. will improve it. https://www.tensorflow.org/api_docs/python/tf/IndexedSlices

eric-haibin-lin commented on python/mxnet/optimizer.py:
```diff
@@ -342,17 +342,20 @@ def create_state(self, index, weight):
         momentum = None
         weight_master_copy = None
         if self.multi_precision and weight.dtype == numpy.float16:
+            assert(weight.stype == 'default'), \
```
Basically we want to add sparse update support for optimzers one by one. It either requires new kernel for the operator or requires some changes in the python optimizer definition. For now if an old optimizer receives a sparse input, it's going to convert input to dense and executes in the old way. I'll update the document of optimizer to indicate whether it supports sparse update so that people know what to expect.

eric-haibin-lin replied:
I added an assert here because it's not yet tested. Let me add a test and remove this.

piiswrong commented:
1. what's the use case for more general sparse formats? Scipy doesn't seem to support N-D sparse tensors. Are there other libraries that supports this?
2. The number of sparse formats grow exponentially and the interaction between them (for binary op) grow quadratically. This only make sense if you can generate kernels. Taco seems to do this for cpu but requires compiler toolchain at runtime. Is it feasible to include it in MXNet? Is is feasible to do it for GPU?

junrushao1994 commented on src/executor/graph_executor.h:
```diff
@@ -19,6 +19,8 @@
 #include <vector>
 #include "./exec_pass.h"
 
+#define EXECUTOR_DEBUG 0
```
Delete this line?

eric-haibin-lin replied:
I'll consider change the DEBUG flag to run time environment variable and limit the debug message to only include the storage types of the node entries in the graph. This is something very useful when running a model and check what happens for each operator.

eric-haibin-lin commented on include/mxnet/ndarray.h:
```diff
@@ -93,24 +154,95 @@ class NDArray {
+  /*!
+   * \return the shape of aux data at ith index. If it doesn't exist, return an empty one.
+   */
+  inline const TShape aux_shape(size_t i) const {
```
Sure

jermainewang commented:
I think we agree a more general sparse scheme is good. The question is how much efforts it costs and how important is it. I think the change can be limited to only the data structure. It is fine our hand-written code only targets specific format combination.
I think having Taco in MXNet is great. I am not sure about the GPU story. Does TF support sparse operation on GPU?

eric-haibin-lin commented:
@jermainewang  Both PyTorch and Caffe2 have sparse feature on GPU, but looks like tacoers haven't worked on GPU yet.

eric-haibin-lin commented:
Hi everyone,

I've updated the branch to include code changes to address comments from reviewers (except the discussion re: taco's more general storage format). 

Recent changes also include:
- added sparse ndarray classes to python api markdown
- added support storage fallback with mutable inputs 
- updated documents for sparse ndarray classes 
- removed todense(), to_rsp(), to_csr() and replaced with tostype()
- changed executor debug str from macro to env var

Known issues to fix include:
- storage fallback fails on batch norm operator https://github.com/apache/incubator-mxnet/issues/7346

I should acknowledge that this PR contains code from @reminisce(sparse dot, sparse retain, cast_storage, sparse elemwise sum, ndarray and infer_storage refactor and many other things) and @stefanhenneking (many optimized kernels for gpu sparse dot and cast_storage operator). This project won't go this far without their significant effort.

piiswrong commented on python/mxnet/ndarray/sparse_ndarray.py:
```diff
@@ -0,0 +1,906 @@
+    if int(_os.environ.get("MXNET_ENABLE_CYTHON", True)) == 0:
+        from .._ctypes.ndarray import NDArrayBase, _set_ndarray_class
+    elif _sys.version_info >= (3, 0):
+        from .._cy3.ndarray import NDArrayBase, _set_ndarray_class
```
why not import these from ndarray to void try except?

formath commented on example/sparse/linear_classification.py:
```diff
@@ -78,18 +78,18 @@ def next(self):
      norm_init = mx.initializer.Normal(sigma=0.01)
-     v = mx.symbol.Variable("v", shape=(feature_dim, 1), init=norm_init, stype='row_sparse')
-     embed = mx.symbol.dot(x, v)
+     weight = mx.symbol.Variable("weight", shape=(feature_dim, 1), init=norm_init, stype='row_sparse')
```
Is `weight` a dense vector of size `feature_dim`? In industrial circumstance, the feature id may be generated by hash, so `feature_dim` could be `numeric_limits<uint64_t>`.

eric-haibin-lin replied:
Yes, we are aware that the dimension could be 64 bit and we're trying to fix that before 1.0 release  https://github.com/apache/incubator-mxnet/issues/7319

CNevd commented on python/mxnet/model.py:
```diff
@@ -113,25 +127,36 @@ def _update_params_on_kvstore(param_arrays, grad_arrays, kvstore, param_names):
         kvstore.push(name, grad_list, priority=-index)
         # pull back the weights
-        kvstore.pull(name, arg_list, priority=-index)
+        if _contains_non_default_storage(arg_list):
```
if arg_list contains both row_sparse and dense weights, it will skip all the arg_list?

eric-haibin-lin replied:
Right. For distributed training we want to reduce the network traffic and use `row_sparse_pull` instead. We may change it if it's not user-friendly. One distributed trainig example with sparse weight is here: https://github.com/eric-haibin-lin/mxnet/blob/sparse/example/sparse/linear_classification.py#L171

CNevd replied:
Can we just skip the row_sparse weights and pull the dense weights normally or any other prettier way to do this?

eric-haibin-lin replied:
Normally given a key, the ndarrays in the arg_list will be either all dense or all sparse, right?

CNevd replied:
Consider arg_list has 1  row_sparse weight and many dense weights, ```_contains_non_default_storage(arg_list)``` will be true, so user have to call  ```kvstore.pull``` for all the dense weights manually out of ```model.py```.  maybe there is some way to split them (row_sparse and dense) so we can only pull row_sparse weights manually?

eric-haibin-lin replied:
If the arg_list does contain what you described, yes, pull is skipped as you said. However, I failed to see why the arg_list for **one variable** will contain a mixture of sparse and non-sparse values if you do data parallel with executor group. I don't know how to interpret why such case (mixture of storage type) will happen. Do you have any use case in mind?
I probably should explicit assert this assumption in the code.

CNevd replied:
Ok got the point, it deals with only one variable

piiswrong commented:
Since this cannot be rebased I'll just squash it into one commit

fredrikbk commented:
Hi @eric-haibin-lin,  @jermainewang, @piiswrong, all,

I'm one of the taco developers. I know that this pull request is closed, but I wanted to add a comment on taco. We're thinking about taco on GPUs, since it is a requirement for entry for at least ML. We think it can be done, but that it's going to be a fair amount of work to make it fast. We'll likely be working on it this spring.

gautamkmr commented:
test_kvstore_gpu.test_row_sparse_pull seems to a bit  flaky and has been failing randomly for egin recent following builds 475, 482. I'll create an issue for this.
